### PR TITLE
Algorithm refactoring

### PR DESCRIPTION
**Pull request by andriiknu**
Main benefits:
1. Analysis algorithm implementation is slightly improved. The Btag mask is built immediately after creating trijet combinations and then applied to the Trijet_p4 column at the defining stage. That's why all following after the Trijet_p4 columns are automatically btagged
2. Avoided all explicit loops

### CODE REVIEW

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thank you @andriiknu ! This PR still needs a little bit of clean-up before review :smile: 

If I understand correctly the changes in commit "Algorithm reimplemented" depend on the changes in commit "Switched to NanoAOD", so I guess we can wait for this latter commit to be merged before we review and merge this one. However, in the meanwhile:
- the commit "added validation script" does not belong in this PR and can be removed
- "Algorithm reimplemented" is perfect as a brief description of the changes: it says "what". Then the longer description (the body of the commit message) needs to explain "why". For example you can put the text you used in the PR description as the body of the commit message (I like https://chris.beams.io/posts/git-commit/ as a good set of rules for good git commit messages)

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:152]:
```suggestion
    # Trijet_btag is a helpful array mask indicating whether or not the maximum btag value in Trijet is larger than the 0.5 threshold
```

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:159]: Thanks to argument-dependent lookup (basically C++ looks for a function also in the namespace of its arguments) this should also work and it's easier on the eyes (same below with other invocations of `Take`, `ArgMax` and `Combinations`):

```suggestion
            auto J1_btagCSVV2 = Take(Jet_btagCSVV2_masked, Trijet[0]);
            auto J2_btagCSVV2 = Take(Jet_btagCSVV2_masked, Trijet[1]);
            auto J3_btagCSVV2 = Take(Jet_btagCSVV2_masked, Trijet[2]);
```

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:176]: no need for these extra newlines:
```suggestion
```

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:172]: I agree this is more readable than the for loop, however 4 Defines are slower than 1 :smile: if we don't need J1, J2 and J3 anywhere else we can just create local variables like you do for `Trijet_btag` I guess?

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:186]: nice!!

**Review by eguiraud — COMMENTED** [review]
Thank you @andriiknu , I left a few comments. I also have a couple more generic ones:

- we should resolve the FIXMEs before merging
- can you confirm histogram contents are unchanged (or anyway agreement with the reference is the same) before and after these changes?

**Comment by andriiknu** [comment]
Produced histograms are exactly the same. I left FIXME just as reminder what should be improved. I didn't mean a bug. Maybe it was better to call it `to do`, but anyway I didn't mean that something is broken

**Comment by eguiraud** [comment]
> Produced histograms are exactly the same as in previous PR

great! thanks for confirming!

> I left FIXME just as reminder what should be improved

I see. Nevertheless, this is a public repository that ROOT users will look at as an example of how an RDF analysis should be written. The implementation should not have FIXMEs or TODOs in it: it shouldn't give the impression that it's unfinished once we tag a v1. Feel free to open GitHub issues if there are to-do items that you don't want to lose track of!

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py:159]: Do you mean I can delete `ROOT::VecOps::` anywhere it is?

Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py:159]: It works, but warning messages appeared about lines 139-140 
```py
        ROOT::VecOps::Construct<ROOT::Math::PxPyPzMVector>(
            ROOT::VecOps::Construct<ROOT::Math::PtEtaPhiMVector>(
```:
input_line_85:4:13: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
            Construct<ROOT::Math::PtEtaPhiMVector>(
            ^
input_line_85:3:9: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
        Construct<ROOT::Math::PxPyPzMVector>(
        ^

Comment (eguiraud) [reply analyses/cms-open-data-ttbar/analysis.py:159]: looks like the function template needs the fully qualified name. let's leave it there just for the `Construct` calls

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [inline analyses/cms-open-data-ttbar/analysis.py:160]: Does the expression have a nice look? A more natural look would have `max(J1_btagCSVV2, J2_btagCSVV2, J3_btagCSVV2)>0.5`, but I haven't seen such a function in ROOT

Comment (eguiraud) [reply analyses/cms-open-data-ttbar/analysis.py:160]: my brain can easily understand the current expression as "at least one of the elements of J1, J2 and J3 should be greater than the btag threshold". I actually find the version with `max` harder to understand. might be subjective though.

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py]: Yes, I've defined `Jn` as separate columns as I needed them for debugging numerical discrepancies. Thank you for pointing it out. I removed them

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:149]: since we are touching this line let's add consistent spacing:
```suggestion
    df = df.Define("Trijet", "Combinations(Jet_pt[Jet_pt_mask], 3)")
```

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py:160]: Okay, in this case, am removing the corresponding `FIXME`

**Comment by andriiknu** [comment]
`FEXME` removed

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:132]: sorry, but since we are touching this line let's make it more readable:
```suggestion
    df = df.Filter("Sum(Jet_btagCSVV2_masked > 0.5) > 1")
```

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:149]: I just realized naming of this column is "wrong": this column does not contain a "trijet", it contains indices that to make trijets. what's a better name? "trijet_idx" maybe?

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:234]: naming: "_masked" is not the most informative. maybe "Jet_pt_ptcut" (it's also one character shorter)? or "Jet_pt_highpt"?

File: analyses/cms-open-data-ttbar/analysis.py
Comment (eguiraud) [inline analyses/cms-open-data-ttbar/analysis.py:256]: I know, I know, I'm nitpicking, but since we are here... :)

```suggestion
    df4j1b = df.Filter("Sum(Jet_btagCSVV2_masked >= 0.5) == 1")\
```

**Review by eguiraud — CHANGES_REQUESTED** [review]
GREAT!!! Just a couple of minor nitpicks left to tie the changes with a nice bow before merging :smile:

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py:234]: "Jet_high_pt"?

Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py]: aa, I've understood, okay

Comment (eguiraud) [reply analyses/cms-open-data-ttbar/analysis.py]: that makes sense for `Jet_pt` but `Jet_btagCSVV2_high` doesn't work. since they are masked the same way I was trying to give them the same suffix. but we can also use two different suffixes (e.g. `_high` for one and `_highpt` for the other)

Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py]: Thanks, I already added `ptcut` for each masked column

File: analyses/cms-open-data-ttbar/analysis.py
Comment (andriiknu) [reply analyses/cms-open-data-ttbar/analysis.py]: Thanks, it is slightly confusing. I've added `_idx` to the `Trijet`

**Review by eguiraud — APPROVED** [review]
LGTM! :rocket:

**Pull request merged**
